File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.js
class Queue {
 constructor() {
   this.items = {};
   this.headIndex = 0;
   this.tailIndex = 0;
 }
 enqueue(item) {
   this.items[this.tailIndex] = item;
   this.tailIndex++;
 }
 dequeue() {
   const item = this.items[this.headIndex];
   delete this.items[this.headIndex];
   this.headIndex++;
   return item;
 }
 peek() {
   return this.items[this.headIndex];
 }
 getLength() {
   return this.tailIndex - this.headIndex;
 }
}

let fs = require("fs");
let input = fs.readFileSync("/dev/stdin").toString().split("\n");

const [n, m, r] = input[0].split(" ").map(Number);
let graph = Array.from({length: n+1}, () => []);
let visited = Array.from({length: n+1}, () => false);
let answer = Array.from({length: n+1}, () => 0);
let order = 1;

for (let i=1; i<m+1; i++) {
  const [u, v] = input[i].split(" ").map(Number);
  graph[u].push(v);
  graph[v].push(u);
}

for (let i=1; i<n+1; i++) {
  graph[i].sort((a, b) => a-b);
}

function bfs(x) {
  q = new Queue();
  q.enqueue(x);
  visited[x] = true;

  while (q.getLength()) {
    cur = q.dequeue();
    visited[cur] = true;
    answer[cur] = order++;

    for (i of graph[cur]) {
      if (!visited[i]) {
        visited[i] = true;
        q.enqueue(i);
      }
    }
  }
}

bfs(r);
// console.log(answer);
console.log(answer.slice(1, ).join("\n"));

File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.py
import sys
from collections import deque

input = sys.stdin.readline

N, M, R = map(int, input().split())

graph = [[] for _ in range(N+1)]
visited = [False for _ in range(N+1)]
answer = [0 for _ in range(N+1)]

for _ in range(M):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

for i in range(N+1):
    graph[i].sort()

count = 1
q = deque()
q.append(R)
visited[R] = True
answer[R] = count
count += 1

while q:
    cur = q.popleft()

    for next in graph[cur]:
        if not visited[next]:
            q.append(next)
            visited[next] = True
            answer[next] = count
            count += 1

for i in range(1, N+1):
    print(answer[i])
